pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -232,16 +233,30 @@ def fetch_remote_url(remote):
     def normalize_url(url):
         return url.rstrip("/") + "/"
 
-    remote_url = normalize_url(remote.url)
-    if get_repomd_file(remote, remote_url):
-        return remote_url
+    try:
+        normalized_remote_url = normalize_url(remote.url)
+        get_repomd_file(remote, normalized_remote_url)
+        # just check if the metadata exists
+        return normalized_remote_url
+    except ClientResponseError as exc:
+        log.info(
+            _("Attempting to resolve a true url from potential mirrolist url '{}'").format(
+                remote.url
+            )
+        )
+        remote_url = fetch_mirror(remote)
+        if remote_url:
+            log.info(
+                _("Using url '{}' from mirrorlist in place of the provided url {}").format(
+                    remote_url, remote.url
+                )
+            )
+            return normalize_url(remote_url)
 
-    remote_url = fetch_mirror(remote)
-    if remote_url:
-        log.info(_("Using url '{}' from mirrorlist").format(remote_url))
-        return normalize_url(remote_url)
+        if exc.status == 404:
+            raise ValueError(_("An invalid remote URL was provided: {}").format(remote.url))
 
-    raise ValueError(_("An invalid remote URL was provided: {}").format(remote.url))
+        raise exc
```

Comment:
Can we tell the difference between "this actually was **not** a mirrorlist URL" and "it was a mirrorlist, but we threw exceptions for every mirror we tried because The Internet Fell Over"? As the code stands, if we can't find repodata/repomd.xml, we'll then treat the remote-url as a mirrorlist and try each mirror in turn. If they all fail to find a repomd.xml, no matter the reason, fetch_mirror() returns None, and if the initial error wasn't a 404, we raise the initial error.

Hm. That's pretty much where we are today. I had to write it all down to see it. Now that I've written it down - do you see any other hole in this logic?

Reply:
That's a pretty accurate summary of what it's doing, I think.  @jlsherrill do you have any input